### ISSUE TITLE
Handle ReadonlyArray

### ISSUE THREAD

**Issue by ellis** [utatti/lens.ts#5]
Does lens.ts handle ReadonlyArray?  If not, this is a feature request for such support.  Here's an example that won't compile:

```typescript
type Person = {
	readonly name: string;
	readonly age: number;
	readonly accounts: ReadonlyArray<Account>;
};

type Account = {
	readonly type: string;
	readonly handle: string;
};

const personL = lens<Person>();
personL.accounts.i(0)
```

Here's the error message from typescript:

```
[ts] Cannot invoke an expression whose type lacks a call signature. Type 'never' has no compatible call signatures.
(property) accounts: LensImpl<{
    readonly name: string;
    readonly age: number;
    readonly accounts: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>;
}, ReadonlyArray<{
    readonly type: string;
    readonly handle: string;
}>> & ArrayLens<{
    readonly name: string;
    readonly age: number;
    readonly accounts: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>;
}, ReadonlyArray<{
    readonly type: string;
    readonly handle: string;
}>> & {
    readonly includes: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (searchElement: {
        readonly type: string;
        readonly handle: string;
    }, fromIndex?: number | undefined) => boolean> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (searchElement: {
        readonly type: string;
        readonly handle: string;
    }, fromIndex?: number | undefined) => boolean> & ObjectLens;
    readonly length: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, number> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, number> & number;
    readonly toString: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, () => string> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, () => string> & ObjectLens;
    readonly toLocaleString: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, () => string> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, () => string> & ObjectLens;
    readonly concat: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, {
        (...items: ConcatArray<{
            readonly type: string;
            readonly handle: string;
        }>[]): {
            readonly type: string;
            readonly handle: string;
        }[];
        (...items: ({
            readonly type: string;
            readonly handle: string;
        } | ConcatArray<{
            readonly type: string;
            readonly handle: string;
        }>)[]): {
            readonly type: string;
            readonly handle: string;
        }[];
    }> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, {
        (...items: ConcatArray<{
            readonly type: string;
            readonly handle: string;
        }>[]): {
            readonly type: string;
            readonly handle: string;
        }[];
        (...items: ({
            readonly type: string;
            readonly handle: string;
        } | ConcatArray<{
            readonly type: string;
            readonly handle: string;
        }>)[]): {
            readonly type: string;
            readonly handle: string;
        }[];
    }> & ObjectLens;
    readonly join: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (separator?: string | undefined) => string> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (separator?: string | undefined) => string> & ObjectLens;
    readonly slice: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (start?: number | undefined, end?: number | undefined) => {
        readonly type: string;
        readonly handle: string;
    }[]> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (start?: number | undefined, end?: number | undefined) => {
        readonly type: string;
        readonly handle: string;
    }[]> & ObjectLens;
    readonly indexOf: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (searchElement: {
        readonly type: string;
        readonly handle: string;
    }, fromIndex?: number | undefined) => number> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (searchElement: {
        readonly type: string;
        readonly handle: string;
    }, fromIndex?: number | undefined) => number> & ObjectLens;
    readonly lastIndexOf: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (searchElement: {
        readonly type: string;
        readonly handle: string;
    }, fromIndex?: number | undefined) => number> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (searchElement: {
        readonly type: string;
        readonly handle: string;
    }, fromIndex?: number | undefined) => number> & ObjectLens;
    readonly every: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (callbackfn: (value: {
        readonly type: string;
        readonly handle: string;
    }, index: number, array: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>) => boolean, thisArg?: any) => boolean> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (callbackfn: (value: {
        readonly type: string;
        readonly handle: string;
    }, index: number, array: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>) => boolean, thisArg?: any) => boolean> & ObjectLens;
    readonly some: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (callbackfn: (value: {
        readonly type: string;
        readonly handle: string;
    }, index: number, array: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>) => boolean, thisArg?: any) => boolean> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (callbackfn: (value: {
        readonly type: string;
        readonly handle: string;
    }, index: number, array: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>) => boolean, thisArg?: any) => boolean> & ObjectLens;
    readonly forEach: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (callbackfn: (value: {
        readonly type: string;
        readonly handle: string;
    }, index: number, array: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>) => void, thisArg?: any) => void> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (callbackfn: (value: {
        readonly type: string;
        readonly handle: string;
    }, index: number, array: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>) => void, thisArg?: any) => void> & ObjectLens;
    readonly map: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, <U>(callbackfn: (value: {
        readonly type: string;
        readonly handle: string;
    }, index: number, array: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>) => U, thisArg?: any) => U[]> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, <U>(callbackfn: (value: {
        readonly type: string;
        readonly handle: string;
    }, index: number, array: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>) => U, thisArg?: any) => U[]> & ObjectLens;
    readonly filter: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, {
        <S extends {
            readonly type: string;
            readonly handle: string;
        }>(callbackfn: (value: {
            readonly type: string;
            readonly handle: string;
        }, index: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => value is S, thisArg?: any): S[];
        (callbackfn: (value: {
            readonly type: string;
            readonly handle: string;
        }, index: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => any, thisArg?: any): {
            readonly type: string;
            readonly handle: string;
        }[];
    }> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, {
        <S extends {
            readonly type: string;
            readonly handle: string;
        }>(callbackfn: (value: {
            readonly type: string;
            readonly handle: string;
        }, index: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => value is S, thisArg?: any): S[];
        (callbackfn: (value: {
            readonly type: string;
            readonly handle: string;
        }, index: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => any, thisArg?: any): {
            readonly type: string;
            readonly handle: string;
        }[];
    }> & ObjectLens;
    readonly reduce: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, {
        (callbackfn: (previousValue: {
            readonly type: string;
            readonly handle: string;
        }, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => {
            readonly type: string;
            readonly handle: string;
        }): {
            readonly type: string;
            readonly handle: string;
        };
        (callbackfn: (previousValue: {
            readonly type: string;
            readonly handle: string;
        }, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => {
            readonly type: string;
            readonly handle: string;
        }, initialValue: {
            readonly type: string;
            readonly handle: string;
        }): {
            readonly type: string;
            readonly handle: string;
        };
        <U>(callbackfn: (previousValue: U, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => U, initialValue: U): U;
    }> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, {
        (callbackfn: (previousValue: {
            readonly type: string;
            readonly handle: string;
        }, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => {
            readonly type: string;
            readonly handle: string;
        }): {
            readonly type: string;
            readonly handle: string;
        };
        (callbackfn: (previousValue: {
            readonly type: string;
            readonly handle: string;
        }, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => {
            readonly type: string;
            readonly handle: string;
        }, initialValue: {
            readonly type: string;
            readonly handle: string;
        }): {
            readonly type: string;
            readonly handle: string;
        };
        <U>(callbackfn: (previousValue: U, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => U, initialValue: U): U;
    }> & ObjectLens;
    readonly reduceRight: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, {
        (callbackfn: (previousValue: {
            readonly type: string;
            readonly handle: string;
        }, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => {
            readonly type: string;
            readonly handle: string;
        }): {
            readonly type: string;
            readonly handle: string;
        };
        (callbackfn: (previousValue: {
            readonly type: string;
            readonly handle: string;
        }, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => {
            readonly type: string;
            readonly handle: string;
        }, initialValue: {
            readonly type: string;
            readonly handle: string;
        }): {
            readonly type: string;
            readonly handle: string;
        };
        <U>(callbackfn: (previousValue: U, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => U, initialValue: U): U;
    }> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, {
        (callbackfn: (previousValue: {
            readonly type: string;
            readonly handle: string;
        }, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => {
            readonly type: string;
            readonly handle: string;
        }): {
            readonly type: string;
            readonly handle: string;
        };
        (callbackfn: (previousValue: {
            readonly type: string;
            readonly handle: string;
        }, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => {
            readonly type: string;
            readonly handle: string;
        }, initialValue: {
            readonly type: string;
            readonly handle: string;
        }): {
            readonly type: string;
            readonly handle: string;
        };
        <U>(callbackfn: (previousValue: U, currentValue: {
            readonly type: string;
            readonly handle: string;
        }, currentIndex: number, array: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => U, initialValue: U): U;
    }> & ObjectLens;
    readonly entries: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, () => IterableIterator<[number, {
            readonly type: string;
            readonly handle: string;
        }]>> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, () => IterableIterator<[number, {
            readonly type: string;
            readonly handle: string;
        }]>> & ObjectLens;
    readonly keys: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, () => IterableIterator<number>> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, () => IterableIterator<number>> & ObjectLens;
    readonly values: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, () => IterableIterator<{
        readonly type: string;
        readonly handle: string;
    }>> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, () => IterableIterator<{
        readonly type: string;
        readonly handle: string;
    }>> & ObjectLens;
    readonly find: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, {
        <S extends {
            readonly type: string;
            readonly handle: string;
        }>(predicate: (this: void, value: {
            readonly type: string;
            readonly handle: string;
        }, index: number, obj: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => value is S, thisArg?: any): S | undefined;
        (predicate: (value: {
            readonly type: string;
            readonly handle: string;
        }, index: number, obj: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => boolean, thisArg?: any): {
            readonly type: string;
            readonly handle: string;
        } | undefined;
    }> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, {
        <S extends {
            readonly type: string;
            readonly handle: string;
        }>(predicate: (this: void, value: {
            readonly type: string;
            readonly handle: string;
        }, index: number, obj: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => value is S, thisArg?: any): S | undefined;
        (predicate: (value: {
            readonly type: string;
            readonly handle: string;
        }, index: number, obj: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>) => boolean, thisArg?: any): {
            readonly type: string;
            readonly handle: string;
        } | undefined;
    }> & ObjectLens;
    readonly findIndex: LensImpl<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (predicate: (value: {
        readonly type: string;
        readonly handle: string;
    }, index: number, obj: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>) => boolean, thisArg?: any) => number> & ArrayLens<{
        readonly name: string;
        readonly age: number;
        readonly accounts: ReadonlyArray<{
            readonly type: string;
            readonly handle: string;
        }>;
    }, (predicate: (value: {
        readonly type: string;
        readonly handle: string;
    }, index: number, obj: ReadonlyArray<{
        readonly type: string;
        readonly handle: string;
    }>) => boolean, thisArg?: any) => number> & ObjectLens;
}
```

**Comment by arturaz** [comment]
I've played with this and it's pretty easy to fix. Just change:

```
export type Unwrap<T> = T extends Array<infer U> ? U : never;
export interface ArrayLens<T, U> extends Array<Lens<T, Unwrap<U>>> {};
public i!: U extends Array<infer E> ? (idx: number) => Lens<T, E> : never;
```
to
```
export type Unwrap<T> = T extends ReadonlyArray<infer U> ? U : never;
export interface ArrayLens<T, U> extends ReadonlyArray<Lens<T, Unwrap<U>>> {};
public i!: U extends ReadonlyArray<infer E> ? (idx: number) => Lens<T, E> : never;
```

As readonlyness only exists as compiler thing, in runtime readonly arrays are actually mutable and all the existing code works.

**Comment by ellis** [comment]
@arturaz Thanks!  I'll try that out.

**Comment by utatti** [comment]
Fixed in 0.5.1. Thanks for your report!
